src/charm.py

```diff
@@ -1,37 +1,292 @@
+    """
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._logging = LogForwarder(self, relation_name="logging")
+        self.tls_certificates = TLSCertificatesProvidesV4(self, CERTIFICATES_RELATION_NAME)
+        self.cert_transfer = CertificateTransferProvides(self, CA_TRANSFER_RELATION_NAME)
+
+        [
+            self.framework.observe(event, self._configure)
+            for event in [
+                self.on[CA_TRANSFER_RELATION_NAME].relation_joined,
+                self.on[CERTIFICATES_RELATION_NAME].relation_changed,
+                self.on.secret_changed,
+                self.on.config_changed,
+                self.on.update_status,
+            ]
+        ]
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_status)
+
+        self._plugin = str(self.model.config.get("plugin", ""))
+
+    def _on_collect_status(self, event: CollectStatusEvent) -> None:
+        """Handle the collect status event."""
+        if not self.unit.is_leader():
+            event.add_status(
+                BlockedStatus(
+                    "this charm does not scale, only the leader unit manages certificates."
+                )
+            )
+            return
+        if err := self._validate_charm_config_options():
+            event.add_status(BlockedStatus(err))
+            return
+        event.add_status(ActiveStatus(self._get_certificate_fulfillment_status()))
+
+    def _configure(self, event: EventBase) -> None:
+        """Configure the Lego provider."""
+        if not self.unit.is_leader():
+            logger.error("only the leader unit can handle certificate requests")
+            return
+        if err := self._validate_charm_config_options():
+            logger.error(err)
+            return
+        self._configure_certificates()
+        self._configure_ca_certificates()
+
+    def _configure_certificates(self):
+        """Attempt to fulfill all certificate requests."""
+        certificate_requests = self.tls_certificates.get_certificate_requests()
+        provided_certificates = self.tls_certificates.get_provider_certificates()
+        certificate_pair_map = {
+            csr: list(
+                filter(
+                    lambda x: x.relation_id == csr.relation_id
+                    and x.certificate_signing_request.raw == csr.certificate_signing_request.raw,
+                    provided_certificates,
+                )
+            )
+            for csr in certificate_requests
+        }
+        with self.maintenance_status("processing certificate requests"):
+            for certificate_request, assigned_certificates in certificate_pair_map.items():
+                if not assigned_certificates:
+                    self._generate_signed_certificate(
+                        csr=certificate_request.certificate_signing_request,
+                        relation_id=certificate_request.relation_id,
+                    )
```

Comment:
Why go to maintenance status every x amount of time, here in most cases there won't be any certificates to generate but the charm will go to that state. We may want to be more precise and only wrap the `_generate_signed_certificate`